File: plc4j/transports/test/src/main/java/org/apache/plc4x/java/transport/test/TestTransportInstance.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.plc4x.java.transport.test;

import org.apache.plc4x.java.spi.transports.api.BaseTransportInstance;
import org.apache.plc4x.java.spi.transports.api.RingBuffer;
import org.apache.plc4x.java.spi.transports.api.TransportInstance;
import org.apache.plc4x.java.spi.transports.api.exceptions.TransportException;
import org.apache.plc4x.java.spi.utils.StaticHelper;
import org.apache.plc4x.java.transport.test.config.TestTransportConfiguration;
import org.apache.plc4x.java.utils.auditlog.api.AuditLog;
import org.apache.plc4x.java.utils.auditlog.api.AuditLogEventType;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

/**
 * Test transport instance implementation using in-memory byte streams.
 * This implementation is useful for testing drivers without actual network connections.
 * <p>
 * Key features:
 * - Thread-safe read/write operations
 * - In-memory ring buffer for data storage
 * - Methods to inject test data and retrieve written data
 */
public class TestTransportInstance extends BaseTransportInstance<TestTransportConfiguration>
        implements TransportInstance<TestTransportConfiguration> {

    private static final Logger LOGGER = LoggerFactory.getLogger(TestTransportInstance.class);
    private static final byte[] EMPTY_BYTES = new byte[0];

    private final RingBuffer readBuffer;
    private final Lock readLock = new ReentrantLock();
    private final Lock writeLock = new ReentrantLock();
    private volatile boolean open = true;

    // For testing: store written data separately
    private final RingBuffer writeBuffer;

    public TestTransportInstance(TestTransportConfiguration configuration, AuditLog auditLog) throws TransportException {
        super(configuration, auditLog);
        this.readBuffer = new RingBuffer(configuration.receiveBufferSize);
        this.writeBuffer = new RingBuffer(configuration.receiveBufferSize);

        LOGGER.debug("Test transport instance created");
        auditLog.write(AuditLogEventType.CONNECT,
            "Test transport instance created");
    }

    /**
     * Injects test data into the read buffer.
     * This simulates receiving data from a remote endpoint.
     *
     * @param data the data to inject
     * @return number of bytes actually written to the buffer
     */
    public int injectTestData(byte[] data) {
        if (data == null || data.length == 0) {
            return 0;
        }

        readLock.lock();
        try {
            int written = readBuffer.write(data);
            LOGGER.debug("Injected {} bytes of test data", written);
            return written;
        } finally {
            readLock.unlock();
        }
    }

    /**
     * Retrieves data that was written to the transport.
     * This is useful for verifying what the driver sent.
     *
     * @param numBytes number of bytes to retrieve
     * @return the retrieved data
     */
    public byte[] getWrittenData(int numBytes) {
        writeLock.lock();
        try {
            if (writeBuffer.availableForReading() < numBytes) {
                return writeBuffer.read(writeBuffer.availableForReading());
            }
            return writeBuffer.read(numBytes);
        } finally {
            writeLock.unlock();
        }
    }

    /**
     * Retrieves all data that was written to the transport.
     *
     * @return all written data
     */
    public byte[] getAllWrittenData() {
        writeLock.lock();
        try {
            return writeBuffer.read(writeBuffer.availableForReading());
        } finally {
            writeLock.unlock();
        }
    }

    /**
     * Waits for the expected number of bytes to be written, then retrieves them.
     * This is useful for test synchronization where we need to wait for the driver to send data.
     *
     * @param expectedBytes the number of bytes to wait for
     * @param timeoutMs     timeout in milliseconds
     * @return the written data
     * @throws TransportException if timeout occurs before expected bytes are available
     */
    public byte[] waitForWrittenData(int expectedBytes, long timeoutMs) throws TransportException {
        long startTime = System.currentTimeMillis();
        long deadline = startTime + timeoutMs;

        while (System.currentTimeMillis() < deadline) {
            writeLock.lock();
            try {
                int available = writeBuffer.availableForReading();
                if (available >= expectedBytes) {
                    byte[] data = writeBuffer.read(expectedBytes);
                    LOGGER.debug("Retrieved {} bytes of written data after {}ms",
                        expectedBytes, System.currentTimeMillis() - startTime);
                    return data;
                }
            } finally {
                writeLock.unlock();
            }

            // Short sleep to avoid busy-waiting
            try {
                Thread.sleep(1);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                throw new TransportException("Interrupted while waiting for written data");
            }
        }

        writeLock.lock();
        try {
            int available = writeBuffer.availableForReading();
            throw new TransportException(String.format(
                "Timeout waiting for written data: expected %d bytes but only %d available after %dms",
                expectedBytes, available, timeoutMs));
        } finally {
            writeLock.unlock();
        }
    }

    /**
     * Returns the number of bytes available in the write buffer.
     *
     * @return number of bytes written
     */
    public int getNumBytesWritten() {
        writeLock.lock();
        try {
            return writeBuffer.availableForReading();
        } finally {
            writeLock.unlock();
        }
    }

    @Override
    public boolean isOpen() {
        return open;
    }

    @Override
    public int getNumBytesAvailable() throws TransportException {
        readLock.lock();
        try {
            if (!isOpen()) {
                return 0;
            }
            return readBuffer.availableForReading();
        } finally {
            readLock.unlock();
        }
    }

    @Override
    public byte[] peekReadableBytes(int numBytes) throws TransportException {
        if (numBytes <= 0) {
            return EMPTY_BYTES;
        }

        readLock.lock();
        try {
            ensureOpen();

            if (readBuffer.availableForReading() < numBytes) {
                throw new TransportException(
                    String.format("Requested %d bytes but only %d available", numBytes, readBuffer.availableForReading())
                );
            }

            return readBuffer.peek(numBytes);
        } catch (TransportException e) {
            getAuditLog().write(AuditLogEventType.ERROR, "Error in peekReadableBytes: " + e.getMessage());
            throw e;
        } finally {
            readLock.unlock();
        }
    }

    @Override
    public byte[] read(int numBytes) throws TransportException {
        if (numBytes <= 0) {
            return EMPTY_BYTES;
        }

        readLock.lock();
        try {
            ensureOpen();

            if (readBuffer.availableForReading() < numBytes) {
                throw new TransportException(
                    String.format("Requested %d bytes but only %d available", numBytes, readBuffer.availableForReading())
                );
            }

            byte[] bytes = readBuffer.read(numBytes);

            if (getAuditLog().isEnabled()) {
                getAuditLog().write(AuditLogEventType.OUTGOING_BYTES, "Read: " + StaticHelper.ENCODE_HEX(bytes));
            }

            return bytes;
        } catch (TransportException e) {
            getAuditLog().write(AuditLogEventType.ERROR, "Error in read: " + e.getMessage());
            throw e;
        } finally {
            readLock.unlock();
        }
    }

    @Override
    public void write(byte[] bytes) throws TransportException {
        if (bytes == null || bytes.length == 0) {
            return;
        }

        writeLock.lock();
        try {
            ensureOpen();

            int written = writeBuffer.write(bytes);
            if (written < bytes.length) {
                throw new TransportException(
                    String.format("Could only write %d of %d bytes", written, bytes.length)
                );
            }

            if (getAuditLog().isEnabled()) {
                getAuditLog().write(AuditLogEventType.OUTGOING_BYTES, "Write: " + StaticHelper.ENCODE_HEX(bytes));
            }
        } catch (TransportException e) {
            getAuditLog().write(AuditLogEventType.ERROR, "Error in write: " + e.getMessage());
            throw e;
        } finally {
            writeLock.unlock();
        }
    }

    @Override
    public void close() throws TransportException {
        if (!open) {
            return;
        }

        writeLock.lock();
        try {
            readLock.lock();
            try {
                open = false;
                LOGGER.debug("Test transport closed");
                getAuditLog().write(AuditLogEventType.CLOSE, "Closed");
            } finally {
                readLock.unlock();
            }
        } finally {
            writeLock.unlock();
        }
    }

    /**
     * Ensures the connection is still open, throws exception otherwise.
     */
    private void ensureOpen() throws TransportException {
        if (!isOpen()) {
            throw new TransportException("Transport is closed");
        }
    }

}

File: plc4net/transports/test/TestTransportInstance.cs
//
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//      https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.
//

using System;
using System.Threading;
using org.apache.plc4net.spi.transports;

namespace org.apache.plc4net.transports.test
{
    /// <summary>
    /// In-memory transport instance for testing. Data flows in two directions:
    /// the test harness <see cref="InjectTestData"/> into the read buffer (simulating
    /// bytes arriving from the wire), and the driver writes into the write buffer
    /// (from which the harness reads with <see cref="GetWrittenData"/>).
    /// </summary>
    /// <remarks>
    /// Mirrors the Java SPI3 <c>TestTransportInstance</c>. Thread-safe: the read
    /// and write paths each have their own lock, and tests can inject data from a
    /// background thread while the driver reads on its own.
    /// </remarks>
    public class TestTransportInstance : BaseTransportInstance
    {
        private readonly RingBuffer _readBuffer;
        private readonly RingBuffer _writeBuffer;
        private readonly object _readLock = new object();
        private readonly object _writeLock = new object();
        private volatile bool _open = true;

        /// <summary>
        /// Creates an instance backed by the given configuration.
        /// </summary>
        public TestTransportInstance(TestTransportConfiguration configuration)
            : base(configuration)
        {
            _readBuffer = new RingBuffer(configuration.ReceiveBufferSize);
            _writeBuffer = new RingBuffer(configuration.ReceiveBufferSize);
        }

        // ── test-facing API ──────────────────────────────────────────

        /// <summary>
        /// Feeds bytes into the read buffer, simulating data arriving from the
        /// remote endpoint. The transport's data listener (if any) is notified.
        /// </summary>
        /// <returns>The number of bytes actually written.</returns>
        public int InjectTestData(byte[] data)
        {
            if (data == null || data.Length == 0)
            {
                return 0;
            }

            int written;
            lock (_readLock)
            {
                written = Math.Min(data.Length, _readBuffer.RemainingForWriting);
                _readBuffer.Write(data, 0, written);
            }

            return written;
        }

        /// <summary>
        /// Returns up to <paramref name="numBytes"/> bytes that the driver has
        /// written to the transport. Consumes them.
        /// </summary>
        public byte[] GetWrittenData(int numBytes)
        {
            lock (_writeLock)
            {
                var available = _writeBuffer.AvailableForReading;
                var toRead = Math.Min(numBytes, available);
                if (toRead <= 0)
                {
                    return Array.Empty<byte>();
                }
                return _writeBuffer.Read(toRead);
            }
        }

        /// <summary>Returns and consumes all bytes the driver has written.</summary>
        public byte[] GetAllWrittenData()
        {
            lock (_writeLock)
            {
                var available = _writeBuffer.AvailableForReading;
                return available <= 0
                    ? Array.Empty<byte>()
                    : _writeBuffer.Read(available);
            }
        }

        /// <summary>
        /// Returns the count of unread bytes the driver has written.
        /// </summary>
        public int GetNumBytesWritten()
        {
            lock (_writeLock)
            {
                return _writeBuffer.AvailableForReading;
            }
        }

        /// <summary>
        /// Waits for at least <paramref name="expectedBytes"/> to be written,
        /// then returns them. Times out with a <see cref="TransportException"/>
        /// if the deadline passes without enough data.
        /// </summary>
        public byte[] WaitForWrittenData(int expectedBytes, int timeoutMs)
        {
            var deadline = Environment.TickCount64 + timeoutMs;
            lock (_writeLock)
            {
                while (_writeBuffer.AvailableForReading < expectedBytes)
                {
                    var remaining = deadline - Environment.TickCount64;
                    if (remaining <= 0)
                    {
                        throw new TransportException(
                            $"Timeout waiting for {expectedBytes} bytes; only " +
                            $"{_writeBuffer.AvailableForReading} arrived after {timeoutMs} ms.");
                    }

                    // Woken by the Monitor.PulseAll in Write/Close, so a write is
                    // observed the moment it lands. The old form polled every 1 ms
                    // on a thread-pool thread; under parallel-test load on a CI
                    // runner the writer's own continuation could be starved past
                    // the deadline, which is what made the caller flaky.
                    Monitor.Wait(_writeLock, (int)remaining);
                }

                return _writeBuffer.Read(expectedBytes);
            }
        }

        // ── ITransportInstance ────────────────────────────────────────

        public override bool IsOpen => _open;

        public override int GetNumBytesAvailable()
        {
            lock (_readLock)
            {
                return _open ? _readBuffer.AvailableForReading : 0;
            }
        }

        public override byte[] PeekReadableBytes(int numBytes)
        {
            if (numBytes <= 0)
            {
                return Array.Empty<byte>();
            }

            lock (_readLock)
            {
                EnsureOpen();
                if (_readBuffer.AvailableForReading < numBytes)
                {
                    throw new TransportException(
                        $"Requested {numBytes} bytes but only {_readBuffer.AvailableForReading} available.");
                }
                return _readBuffer.Peek(numBytes);
            }
        }

        public override byte[] Read(int numBytes)
        {
            if (numBytes <= 0)
            {
                return Array.Empty<byte>();
            }

            lock (_readLock)
            {
                EnsureOpen();
                if (_readBuffer.AvailableForReading < numBytes)
                {
                    throw new TransportException(
                        $"Requested {numBytes} bytes but only {_readBuffer.AvailableForReading} available.");
                }
                return _readBuffer.Read(numBytes);
            }
        }

        public override void Write(byte[] bytes)
        {
            if (bytes == null || bytes.Length == 0)
            {
                return;
            }

            lock (_writeLock)
            {
                EnsureOpen();
                if (bytes.Length > _writeBuffer.RemainingForWriting)
                {
                    throw new TransportException(
                        $"Cannot write {bytes.Length} bytes; only {_writeBuffer.RemainingForWriting} free.");
                }
                _writeBuffer.Write(bytes);
                Monitor.PulseAll(_writeLock);
            }
        }

        public override void Close()
        {
            if (!_open)
                return;

            // Only hold _writeLock; _readLock callers check _open after acquiring
            // their lock, so setting _open = false under _writeLock is sufficient.
            lock (_writeLock)
            {
                _open = false;
                Monitor.PulseAll(_writeLock);
            }
        }

        }
}
